core/src/state.rs

```diff
@@ -0,0 +1,1905 @@
+//! This module provides the [`State`] — an in-memory representation of the current blockchain state.
+use std::{borrow::Borrow, collections::BTreeSet, marker::PhantomData, sync::Arc, time::Duration};
+
+use eyre::Result;
+use iroha_config::{
+    base::proxy::Builder,
+    state::{Configuration, ConfigurationProxy},
+};
+use iroha_crypto::HashOf;
+use iroha_data_model::{
+    account::AccountId,
+    block::SignedBlock,
+    events::notification::{TriggerCompletedEvent, TriggerCompletedOutcome},
+    isi::error::{InstructionExecutionError as Error, MathError},
+    parameter::Parameter,
+    permission::{PermissionTokenSchema, Permissions},
+    prelude::*,
+    query::error::{FindError, QueryExecutionFail},
+    role::RoleId,
+};
+use iroha_logger::prelude::*;
+use iroha_primitives::small::SmallVec;
+use parking_lot::Mutex;
+use range_bounds::RoleIdByAccountBounds;
+use serde::{
+    de::{DeserializeSeed, MapAccess, Visitor},
+    Deserializer, Serialize,
+};
+use storage::{
+    cell::{Block as CellBlock, Cell, Transaction as CellTransaction, View as CellView},
+    storage::{
+        Block as StorageBlock, RangeIter, Snapshot as StorageSnapshot, Storage,
+        Transaction as StorageTransaction, View as StorageView,
+    },
+};
+
+use crate::{
+    block::CommittedBlock,
+    executor::Executor,
+    kura::Kura,
+    query::store::LiveQueryStoreHandle,
+    role::RoleIdWithOwner,
+    smartcontracts::{
+        triggers::{
+            self,
+            set::{LoadedActionTrait, LoadedWasm, Set as TriggerSet},
+        },
+        wasm, Execute,
+    },
+    tx::TransactionExecutor,
+    Parameters, PeersIds,
+};
+
+/// The global entity consisting of `domains`, `triggers` and etc.
+/// For example registration of domain, will have this as an ISI target.
+#[derive(Default, Serialize)]
+pub struct World {
```

Comment:
What is the reason to let each field be `Cell` or `Storage` instead of having a single `Cell<SomeTransactionalRoot>` ?

Reply:
So `Storage` is `BtreeSet`-like structure which could be used effectively for preserving multiple versions (by updating only subtree of data).

While `Cell` is versioned storage for single entitiy. 

So in case where i expect data to be changed altogether or size isn't very big i use `Cell` if data is key value and expected to be big i use `Storage`.

Reply:
Probably need better names.

Reply:
I said "for example" because the current `World` is not a transactional root

Reply:
> My question was, in other words, could we consider, for example, World as that single entity?
Do we need concread structures for each field of World?

Oh, got you now.

Yes, we need `concread` structures for each field to make changes independent, e.g. when minting assets we only update single account, we don't need clone of the whole world.

If we would use `Cell` world this would be essentially the same as just cloning (which we have right now).